libs/langchain/langchain/vectorstores/neo4j_vector.py

```diff
@@ -0,0 +1,623 @@
+        not supported.
+        """
+        version = self.query("CALL dbms.components()")[0]["versions"][0]
+        version_tuple = tuple(map(int, version.split(".")))
+        target_version = (5, 11, 0)
+
+        if version_tuple < target_version:
+            raise ValueError(
+                "Version index is only supported in Neo4j version 5.11 or greater"
+            )
+
+    def retrieve_existing_index(self) -> Optional[int]:
+        """
+        Check if the vector index exists in the Neo4j database
+        and returns its embedding dimension.
+
+        This method queries the Neo4j database for existing indexes
+        and attempts to retrieve the dimension of the vector index
+        with the specified name. If the index exists, its dimension is returned.
+        If the index doesn't exist, `None` is returned.
+
+        Returns:
+            int or None: The embedding dimension of the existing index if found.
+        """
+
+        index_information = self.query(
+            "SHOW INDEXES YIELD name, type, labelsOrTypes, properties, options "
+            "WHERE type = 'VECTOR' AND name = $index_name "
+            "RETURN name, labelsOrTypes, properties, options ",
+            {"index_name": self.index_name},
+        )
+        try:
+            self.node_label = index_information[0]["labelsOrTypes"][0]
+            self.embedding_node_property = index_information[0]["properties"][0]
+            embedding_dimension = index_information[0]["options"]["indexConfig"][
+                "vector.dimensions"
+            ]
+
+            return embedding_dimension
+        except IndexError:
+            return None
+
+    def create_new_index(self) -> None:
+        """
+        This method constructs a Cypher query and executes it
+        to create a new vector index in Neo4j.
+        """
+        index_query = (
+            "CALL db.index.vector.createNodeIndex("
+            "$index_name,"
+            "$node_label,"
+            "$embedding_node_property,"
+            "toInteger($embedding_dimension),"
+            "$similarity_metric )"
+        )
+
+        parameters = {
+            "index_name": self.index_name,
+            "node_label": self.node_label,
+            "embedding_node_property": self.embedding_node_property,
+            "embedding_dimension": self.embedding_dimension,
+            "similarity_metric": self._distance_strategy,
+        }
+        self.query(index_query, parameters)
+
+    @property
+    def embeddings(self) -> Embeddings:
+        return self.embedding_function
+
+    @classmethod
+    def __from(
+        cls,
+        texts: List[str],
+        embeddings: List[List[float]],
+        embedding: Embeddings,
+        username: str,
+        password: str,
+        url: str,
+        metadatas: Optional[List[dict]] = None,
+        ids: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> Neo4jVector:
+        if ids is None:
```

Comment:
we have two places where we create uuids? so we don't know anymore if the ids were provided from the outside or not, we should probably do this only at the last moment? or inside the create/update statement?

possibly same for metadata, instead of empty maps we can just pass null or no entries